include/ProgramRepresentation/DisjointedPVAliasSets.h

```diff
@@ -0,0 +1,74 @@
+#ifndef DISJOINED_PV_ALIAS_SETS_H
+#define DISJOINED_PV_ALIAS_SETS_H
+
+#include "ProgramRepresentation/PVAliasSet.h"
+
+// manages a disjoint set (https://en.wikipedia.org/wiki/Disjoint-set_data_structure)
+// of program variable must-aliases.
+// each program variable is an element of no more than 1 set.
+class DisjointedPVAliasSets {
+  private:
+    std::list<PVAliasSet> sets;
+
+    // returns a reference to the alias set programVar belongs to. if it's not
+    // found, .end() is returned.
+    typename std::list<PVAliasSet>::iterator findIter(ProgramVariable programVar);
+
+    // puts the elements of set2 into set1 in constant time. set2 is also deleteed
+    // from our set of sets.
+    void merge(typename std::list<PVAliasSet>::iterator set1,
+               typename std::list<PVAliasSet>::iterator set2);
+
+  public:
+    // returns true iff programVar is in any of the disjointed sets
+    bool elementIsInAnySet(ProgramVariable programVar);
+
+    // returns a copy of every disjointed set
+    std::list<PVAliasSet> getSets() const;
+
+    // returns a reference to the alias set that has programVar.
+    // NULL is returned if it's not found.
+    PVAliasSet *getSetRef(ProgramVariable programVar);
+
+    // returns a reference to the alias set that has a program variable
+    // with name cleanedName. NULL is return if it's not found.
+    // it is preferable to use the overload that specifies a ProgramVariable to avoid
+    // collisions between local (%) and global (@) LLVM identifiers. see
+    // LLVM docs: https://llvm.org/docs/LangRef.html#identifiers
+    PVAliasSet *getSetRef(const std::string& cleanedName);
+
+    // returns a reference to the alias set that has a program variable
+    // with value val. NULL is return if it's not found.
+    PVAliasSet *getSetRef(Value* val);
+
+    // unions the sets that contain either elementA or elementB. if elementA == elementB, or
+    // elementA or elementB are not actually elements of any set, the union is not performed.
+    void unionSets(ProgramVariable elementA, ProgramVariable elementB);
+
+    // creates a new disjoint set if it's already not an element in any of the
+    // existing sets.
```

Comment:
What if `programVar` is an element in an existing set? It's unclear from this comment whether:
* a precondition of `makeSet` is that its argument is not an element of any set (i.e., the client should check that condition before calling `makeSet`, and calling `makeSet` when the condition is false is an error
* `makeSet` will do nothing if its argument is already in a set

I'd prefer the latter design, but the wording in this comment needs to be more precise regardless.

Reply:
It should be the latter so I'll update the comment. I also just realized that the code itself did not actually check if `programVar` was an existing member, so I'm also going to update the code to do that (it's a 3 line addition)